opm/autodiff/BlackoilModelEbos.hpp

```diff
@@ -1032,150 +1033,98 @@ namespace Opm {
-                        const double hydrocarbon = fs.saturation(FluidSystem::oilPhaseIdx).value() + fs.saturation(FluidSystem::gasPhaseIdx).value();
-
-                        if (hcpv[region] != 0) {
-                            fip_.fip[FIPData::FIP_WEIGHTED_PRESSURE][c] = pv[c] * fs.pressure(FluidSystem::oilPhaseIdx).value() * hydrocarbon / hcpv[region];
-                        } else {
-                            fip_.fip[FIPData::FIP_WEIGHTED_PRESSURE][c] = pres[region] / pv[c];
-                        }
+                elemCtx.updatePrimaryStencil(elem);
+                elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
 
-                        values[region][FIPData::FIP_PV] += fip_.fip[FIPData::FIP_PV][c];
-                        values[region][FIPData::FIP_WEIGHTED_PRESSURE] += fip_.fip[FIPData::FIP_WEIGHTED_PRESSURE][c];
+                unsigned cellIdx = elemCtx.globalSpaceIndex(/*spaceIdx=*/0, /*timeIdx=*/0);
+                const int region = fipnum[cellIdx] - 1;
+                if (region != -1) {
+                    const auto& intQuants = elemCtx.intensiveQuantities(/*spaceIdx=*/0, /*timeIdx=*/0);
+                    const auto& fs = intQuants.fluidState();
+                    const double pv =
+                        ebosSimulator_.model().dofTotalVolume(cellIdx)
+                        * ebosSimulator_.problem().porosity(cellIdx);
+                    fip_.fip[FIPData::FIP_PV][cellIdx] = pv;
+                    const double hydrocarbon = fs.saturation(FluidSystem::oilPhaseIdx).value() + fs.saturation(FluidSystem::gasPhaseIdx).value();
+
+                    //Compute hydrocarbon pore volume weighted average pressure.
+                    //If we have no hydrocarbon in region, use pore volume weighted average pressure instead
+                    if (hcpv[region] != 0.0) {
+                        fip_.fip[FIPData::FIP_WEIGHTED_PRESSURE][cellIdx] = pv * fs.pressure(FluidSystem::oilPhaseIdx).value() * hydrocarbon / hcpv[region];
+                    } else {
+                        fip_.fip[FIPData::FIP_WEIGHTED_PRESSURE][cellIdx] = pres[region] / pv;
                     }
-                }
 
-                // For the frankenstein branch we hopefully can turn values into a vanilla
-                // std::vector<double>, use some index magic above, use one communication
-                // to sum up the vector entries instead of looping over the regions.
-                for(int reg=0; reg < dims; ++reg)
-                {
-                    comm.sum(values[reg].data(), values[reg].size());
+                    values[region][FIPData::FIP_PV] += fip_.fip[FIPData::FIP_PV][cellIdx];
+                    values[region][FIPData::FIP_WEIGHTED_PRESSURE] += fip_.fip[FIPData::FIP_WEIGHTED_PRESSURE][cellIdx];
                 }
-#else
-                // This should never happen!
-                OPM_THROW(std::logic_error, "HAVE_MPI should be defined if we are running in parallel");
```

Comment:
it not necessary (the communicator is always used) and we would have to go through hoops to keep it (which is a weird thing to do for code which is never going to be executed IMO).